modules/nlu/src/backend/engine.ts

```diff
@@ -0,0 +1,112 @@
+import retry from 'bluebird-retry'
+import * as sdk from 'botpress/sdk'
+import crypto from 'crypto'
+import fs from 'fs'
+
+import { Config } from '../config'
+
+import { DucklingEntityExtractor } from './pipelines/entities/duckling_extractor'
+import FastTextClassifier from './pipelines/intents/ft_classifier'
+import { FastTextIndentifier } from './pipelines/language/ft_lid'
+import Storage from './storage'
+
+export default class ScopedEngine {
+  public readonly storage: Storage
+  public minConfidence: number
+  public maxConfidence: number
+
+  private readonly intentClassifier: FastTextClassifier
+  private readonly langDetector: LanguageIdentifier
+  private readonly knownEntityExtractor: EntityExtractor
+
+  private retryPolicy = {
+    interval: 100,
+    max_interval: 500,
+    timeout: 5000,
+    max_tries: 3
+  }
+
+  constructor(private logger: sdk.Logger, private botId: string, private readonly config: Config) {
+    this.storage = new Storage(config, this.botId)
+    this.intentClassifier = new FastTextClassifier(this.logger)
+    this.langDetector = new FastTextIndentifier(this.logger)
+    this.knownEntityExtractor = new DucklingEntityExtractor(this.logger)
+  }
+
+  async init(): Promise<void> {
+    this.minConfidence = this.config.minimumConfidence
+    this.maxConfidence = this.config.maximumConfidence
+
+    if (isNaN(this.minConfidence) || this.minConfidence < 0 || this.minConfidence > 1) {
+      this.minConfidence = 0
+    }
+
+    if (isNaN(this.maxConfidence) || this.maxConfidence < 0) {
+      this.maxConfidence = 1
+    }
+
+    if (await this.checkSyncNeeded()) {
+      await this.sync()
+    }
+  }
+
+  public async sync(): Promise<void> {
+    const intents = await this.storage.getIntents()
+    const modelHash = this._getIntentsHash(intents)
+
+    if (await this.storage.modelExists(modelHash)) {
```

Comment:
IMO it would be cleaner to read if you extracted the inner logic in each conditional branch into private methods
```ts
if (await this.storage.modelExists(modelHash)) {
  return this.loadModel(modelHash)
} else {
 return this.trainModel(intents, modelHash)
}